client/device/cloudftd/upgrade.go

```diff
@@ -46,11 +49,52 @@ func (f *ftdUpgradeService) Upgrade(uid string, softwareVersion string) (*FtdDev
 	if err != nil {
 		return nil, err
 	}
-	err = f.validateFtdVersion(ftdDevice, softwareVersion)
+	err = f.validateFtdVersion(ftdDevice, softwareVersionStr)
 	if err != nil {
 		return nil, err
 	}
 
+	tflog.Debug(f.Ctx, "Triggering the FTD device upgrade.")
+	return f.doUpgrade(ftdDevice, softwareVersionStr)
+}
+
+func (f *ftdUpgradeService) doUpgrade(ftdDevice *FtdDevice, softwareVersionStr string) (*FtdDevice, error) {
+	versionOnDevice, err := ftd.NewVersion(ftdDevice.SoftwareVersion)
+	softwareVersion, err := ftd.NewVersion(softwareVersionStr)
+	if versionOnDevice.Equal(softwareVersion) {
```

Comment:
IIf you follow my comment above, you can avoid having to do this version check again